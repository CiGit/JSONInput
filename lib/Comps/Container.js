'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Store = require('../Store');

var _Store2 = _interopRequireDefault(_Store);

var _higherOrder = require('baobab-react/higher-order');

var _SchemaType = require('./SchemaType');

var _SchemaType2 = _interopRequireDefault(_SchemaType);

var _actions = require('../Store/actions');

var actions = _interopRequireWildcard(_actions);

var _customValidator = require('./../Utils/customValidator');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var BranchedSchemaType = (0, _higherOrder.branch)(_SchemaType2.default, {
    cursors: {
        schema: 'schema',
        status: 'status',
        value: 'value'
    }
});

/**
 * Top Component
 */

var Container = function (_React$Component) {
    _inherits(Container, _React$Component);

    function Container(props) {
        _classCallCheck(this, Container);

        var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Container).call(this, props));

        _this.tree = (0, _Store2.default)();
        _this.tree.select('value').set(props.value);
        _this.tree.select('schema').set(props.schema);
        _this.tree.select('value').on('update', function (event) {
            return _this.props.onChange(event.data.currentData);
        });

        // baobab-react optim as actions are bound on each render -> pure
        _this.ACTIONS = {};
        Object.keys(actions).forEach(function (action) {
            _this.ACTIONS[action] = actions[action].bind(_this.tree, _this.tree);
            return;
        });
        return _this;
    }

    _createClass(Container, [{
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
            this.tree.select('value').set(nextProps.value);
            if (this.props.schema !== nextProps.schema) {
                this.tree.select('schema').set(nextProps.schema);
            }
        }
    }, {
        key: 'shouldComponentUpdate',
        value: function shouldComponentUpdate(nextProps) {
            return nextProps.value !== this.tree.get('value') || nextProps.schema !== this.props.schema;
        }
    }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
            this.tree.release();
        }
    }, {
        key: 'getValue',
        value: function getValue() {
            return this.tree.get('value');
        }
    }, {
        key: 'validate',
        value: function validate() {
            var validationResult = (0, _customValidator.validate)(this.tree.get('value'), this.tree.get('schema'), this.ACTIONS.getFormValue());
            var setErrors = this.ACTIONS.setErrors;

            var errorMap = new Map();
            // Collect each error associated with a given path
            validationResult.errors.forEach(function (error) {
                var errors = errorMap.get(error.property) || [];
                errors.push(error.message); // Add new error
                errorMap.set(error.property, errors);
            });
            errorMap.forEach(function (value, key) {
                setErrors(key.split(/\.|\[|\]/).filter(function (x) {
                    return x !== '';
                }).slice(1), value);
            });
            return validationResult.errors;
        }
    }, {
        key: 'render',
        value: function render() {
            var Rooted = (0, _higherOrder.root)(BranchedSchemaType, this.tree);
            return _react2.default.createElement(Rooted, _extends({}, this.props, {
                path: [],
                actions: this.ACTIONS
            }));
        }
    }]);

    return Container;
}(_react2.default.Component);

exports.default = Container;