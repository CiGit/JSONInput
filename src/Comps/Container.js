import React, { PropTypes } from 'react';
import createTree from '../Store';
import { root, branch } from 'baobab-react/higher-order';
import SchemaType from './SchemaType';
import * as actions from '../Store/actions';
import { validate } from './../Utils/customValidator';

const BranchedSchemaType = branch(SchemaType, {
    cursors: {
        schema: 'schema',
        status: 'status',
        value: 'value'
    }
});

/**
 * Top Component
 */
class Container extends React.Component {
    constructor(props) {
        super(props);
        this.tree = createTree();
        this.tree.select('value').set(props.value);
        this.tree.select('schema').set(props.schema);
        this.tree.select('value')
            .on('update', event => this.props.onChange(event.data.currentData));

        // baobab-react optim as actions are bound on each render -> pure
        this.ACTIONS = {};
        Object.keys(actions)
            .forEach(action => {
                this.ACTIONS[action] = actions[action].bind(this.tree, this.tree);
                return;
            });
    }
    componentWillReceiveProps(nextProps) {
        this.tree.select('value').set(nextProps.value);
        if (this.props.schema !== nextProps.schema) {
            this.tree.select('schema').set(nextProps.schema);
        }
    }
    shouldComponentUpdate(nextProps) {
        return nextProps.value !== this.tree.get('value') || nextProps.schema !== this.props.schema;
    }
    componentWillUnmount() {
        this.tree.release();
    }
    getValue() {
        return this.tree.get('value');
    }
    validate() {
        const validationResult = validate(this.tree.get('value'),
            this.tree.get('schema'), this.ACTIONS.getFormValue());
        const { setErrors } = this.ACTIONS;
        const errorMap = new Map();
        // Collect each error associated with a given path
        validationResult.errors.forEach(error => {
            const errors = errorMap.get(error.property) || [];
            errors.push(error.message); // Add new error
            errorMap.set(error.property, errors);
        });
        errorMap.forEach((value, key) => {
            setErrors(key.split(/\.|\[|\]/).filter(x => x !== '').slice(1), value);
        });
        return validationResult.errors;
    }
    render() {
        const Rooted = root(BranchedSchemaType, this.tree);
        return (
            <Rooted
                {...this.props}
                path={[]}
                actions={this.ACTIONS}
            />);
    }
}
Container.propTypes = {
    onChange: PropTypes.func,
    schema: PropTypes.object,
    value: PropTypes.any
};

export default Container;
