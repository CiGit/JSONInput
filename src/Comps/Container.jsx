import React, { PropTypes } from 'react';
import createTree from '../Store';
import { root, branch } from 'baobab-react/higher-order';
import SchemaType from './SchemaType.jsx';
import * as actions from '../Store/actions';
import { validate } from './../Utils/customValidator';

const BranchedSchemaType = branch(SchemaType, {
    cursors: {
        schema: 'schema',
        status: 'status',
        value: 'value'
    } /* ,
    actions: {
        update,
        defaultValue,
        getStatus,
        getErrors
    }*/
});

/**
 * Top Component
 */
class Container extends React.Component {
    constructor(props) {
        super(props);
        this.tree = createTree();
        this.tree.select('schema').set(props.schema);
        this.tree.select('value').set(props.value);
        this.tree.select('value')
            .on('update', event => this.props.onChange(event.data.currentData));

        // baobab-react optim as actions are bound on each render -> pure
        this.ACTIONS = {};
        Object.keys(actions)
            .forEach(action => this.ACTIONS[action] = actions[action].bind(this.tree, this.tree));
    }
    componentWillReceiveProps(nextProps) {
        this.tree.select('schema').set(nextProps.schema);
        this.tree.select('value').set(nextProps.value);
    }
    componentWillUnmount() {
        this.tree.release();
    }
    validate() {
        const validationResult = validate(this.tree.get('value'), this.props.schema);
        const { setErrors } = this.ACTIONS;
        const errorMap = new Map();
        // Collect each error associated with a given path
        validationResult.errors.forEach(error => {
            const errors = errorMap.get(error.property) || [];
            errors.push(error.message); // Add new error
            errorMap.set(error.property, errors);
        });
        errorMap.forEach((value, key) => {
            setErrors(key.split('.').slice(1), value);
        });
        return validationResult.errors;
    }
    render() {
        const Rooted = root(BranchedSchemaType, this.tree);
        return (<Rooted path={ [] }
                        actions={ this.ACTIONS } />);
    }
}
Container.propTypes = {
    onChange: PropTypes.func,
    schema: PropTypes.object,
    value: PropTypes.any
};

export default Container;
