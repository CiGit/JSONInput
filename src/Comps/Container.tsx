import * as React from 'react';
import { root, branch } from 'baobab-react/higher-order';
import createTree from '../Store/index';
import SchemaType from './SchemaType';
import { setErrors } from '../Store/actions';
import validate from './../Utils/customValidator';

import { Schema } from '../../typings/types';

const BranchedSchemaType = branch(
    {
        schema: 'schema',
        status: 'status',
        value: 'value'
    },
    SchemaType
);

export type Props = {
    onChange: (value: {}, errors: {}[]) => void,
    schema: Schema,
    value?: {}
};
function noop() { };
/**
 * Top Component
 */
class Container extends React.Component<Props, undefined> {
    event: boolean;
    static defaultProps = { schema: {} };
    private tree: any;
    private rooted: React.StatelessComponent<{
        onChange: (value: {}, errors?: {}[]) => void,
        path: string[]
    }>;
    props: Props;

    constructor(props: Props) {
        super(props);
        this.tree = createTree();
        this.updateTree(props.value, props.schema);
        this.rooted = root(this.tree, BranchedSchemaType);
    }
    componentDidMount() {
        this.tree
            .select('value')
            .on('update', (event: { data: { currentData: {} } }) => {
                if (this.event) {
                    this.props.onChange(
                        event.data.currentData,
                        validate(
                            event.data.currentData,
                            this.tree.get('schema'),
                            event.data.currentData
                        ).errors
                    )
                }
            }
            );
    }

    componentWillReceiveProps(nextProps: Props) {
        if (
            nextProps.value === this.tree.get('value') &&
            nextProps.schema === this.props.schema
        ) {
            return;
        }
        this.updateTree(nextProps.value, nextProps.schema);
    }
    componentWillUnmount() {
        this.tree.release();
    }
    shouldComponentUpdate() {
        return false;
    }
    getValue() {
        return this.tree.get('value');
    }
    updateTree(value?: {}, schema?: Schema) {
        this.event = false;
        this.tree.set('value', value);
        this.tree.set('schema', schema);
        this.tree.set('status', {});
        this.tree.commit();
        this.event = true;
    }
    validate() {
        const validationResult = validate(
            this.tree.get('value'),
            this.tree.get('schema'),
            this.tree.get('value')
        );
        const errorMap = new Map<string, string[]>();
        // Collect each error associated with a given path
        validationResult.errors.forEach(error => {
            const errors = errorMap.get(error.property) || [];
            errors.push(error.message); // Add new error
            errorMap.set(error.property, errors);
        });
        errorMap.forEach((value, key) => {
            setErrors(
                this.tree,
                key.split(/\.|\[|\]/).filter(x => x !== '').slice(1),
                value
            );
        });
        return validationResult.errors;
    }
    render() {
        const Rooted = this.rooted;
        return <Rooted onChange={noop} path={[]} />;
    }
}

export default Container;
