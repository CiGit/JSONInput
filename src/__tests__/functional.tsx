import * as React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import defaultViews from '../Comps/Views';
import Container, { setDefaultWidgets } from '../index';

describe('Visibility', () => {
  beforeAll(() => {
    setDefaultWidgets(defaultViews as any);
  });
  afterEach(cleanup);
  test('Simple visibility', () => {
    let val: any = 'Hello';
    const container = document.createElement('div');
    render(
      <Container
        schema={{
          type: 'string',
          visible: () => {
            return false;
          },
        }}
        onChange={v => {
          val = v;
        }}
        value={val}
      />,
      { container, baseElement: container },
    );
    expect(container.firstChild).toMatchSnapshot('Nothing');
    expect(val).toBe('Hello');
    val = 'World';
    render(
      <Container
        schema={{
          type: 'string',
          visible: () => {
            return true;
          },
        }}
        onChange={v => {
          val = v;
        }}
        value={val}
      />,
      { container, baseElement: container },
    );
    expect(container.firstChild).toMatchSnapshot('Input visible');
    expect(val).toBe('World');
  });
  test('Toggle visibility', () => {
    let val: any = { val: 'Hello', visible: false };
    const { container } = render(
      <Container
        schema={{
          type: 'object',
          properties: {
            val: {
              type: 'string',
              visible: (val, formVal) => {
                return formVal.visible;
              },
            },
            visible: {
              type: 'boolean',
            },
          },
        }}
        onChange={v => {
          val = v;
        }}
        value={val}
      />,
    );
    const cbx = container.querySelector('input')!;
    fireEvent.change(cbx);
    expect(container.firstChild).toMatchSnapshot('Hello input invisible');
    expect(val).toEqual({ val: 'Hello', visible: false });
    // Toggle visibility
    cbx.checked = true;
    fireEvent.change(cbx);
    expect(container.firstChild).toMatchSnapshot('Hello input visible');
    expect(val).toEqual({ val: 'Hello', visible: true });
  });
  test('Hides throwing visibility function', () => {
    const { container } = render(
      <Container
        schema={{
          type: 'string',
          visible: () => {
            throw Error('...');
          },
        }}
        onChange={() => {}}
      />,
    );
    expect(container.firstChild).toMatchSnapshot('Nothing');
  });
});
describe('Errored', () => {
  beforeAll(() => {
    setDefaultWidgets(defaultViews as any);
  });
  test('"errored" is not a function', () => {
    const conError = spyOn(console, 'error'); // Hide error messages.
    expect(() =>
      render(
        <Container
          schema={{
            type: 'string',
            value: '',
            errored: undefined,
          }}
          onChange={() => {}}
        />,
      ),
    ).toThrowError('"errored" expects a function');
  });
  test('Error shown on change only', () => {
    const errorMessage = "I'm always wrong";
    const { container, getByText } = render(
      <Container
        schema={{
          type: 'string',
          errored: () => errorMessage,
        }}
        value=""
        onChange={() => {}}
      />,
    );
    // there is no error message
    expect(() => getByText(errorMessage)).toThrow();
    // throw a change event
    fireEvent.change(container.querySelector('input'));
    // Error is shown
    expect(() => getByText(errorMessage)).not.toThrow();
  });
  test('Error hides on change', () => {
    const errorMessage = "I'm not Empty";
    const { container, getByText } = render(
      <Container
        schema={{
          type: 'string',
          errored: val => (val ? errorMessage : ''),
        }}
        value=""
        onChange={() => {}}
      />,
    );
    const input = container.querySelector('input');
    expect(() => getByText(errorMessage)).toThrow();
    input.value = 'Something';
    fireEvent.change(input);
    expect(() => getByText(errorMessage)).not.toThrow();
    input.value = '';
    fireEvent.change(input);
    expect(() => getByText(errorMessage)).toThrow();
  });
});
