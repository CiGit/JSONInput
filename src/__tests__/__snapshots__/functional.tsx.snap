// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visibility Hides throwing visibility function: Nothing 1`] = `null`;

exports[`Visibility Simple visibility: Input visible 1`] = `
<div>
  <label
    class="stringField "
  >
    <span
      class="title"
    />
  </label>
  <input
    type="string"
    value="World"
  />
  <span />
  <span />
</div>
`;

exports[`Visibility Simple visibility: Nothing 1`] = `null`;

exports[`Visibility Toggle visibility: Hello input invisible 1`] = `
<div>
  <label
    class="objectField "
  >
    <span
      class="title"
    />
  </label>
  <div>
    <div>
      <label
        class="booleanField "
      >
        <span
          class="title"
        >
          visible
        </span>
      </label>
      <input
        type="checkbox"
        value="false"
      />
      <span />
      <span />
    </div>
  </div>
  <span />
  <span />
</div>
`;

exports[`Visibility Toggle visibility: Hello input visible 1`] = `
<div>
  <label
    class="objectField "
  >
    <span
      class="title"
    />
  </label>
  <div>
    <div>
      <label
        class="stringField "
      >
        <span
          class="title"
        >
          val
        </span>
      </label>
      <input
        type="string"
        value="Hello"
      />
      <span />
      <span />
    </div>
    <div>
      <label
        class="booleanField "
      >
        <span
          class="title"
        >
          visible
        </span>
      </label>
      <input
        type="checkbox"
        value="true"
      />
      <span />
      <span />
    </div>
  </div>
  <span />
  <span />
</div>
`;
